core/src/main/scala/org/apache/spark/util/ShutdownHookManager.scala

```diff
@@ -232,22 +232,23 @@ private [util] class SparkShutdownHookManager {
     }
   }
 
-  def runAll(): Unit = synchronized {
+  def runAll(): Unit = {
     shuttingDown = true
-    while (!hooks.isEmpty()) {
-      Try(Utils.logUncaughtExceptions(hooks.poll().run()))
+    var nextHook: SparkShutdownHook = null
+    while ({nextHook = hooks synchronized { hooks.poll() }; nextHook != null}) {
+      Try(Utils.logUncaughtExceptions(nextHook.run()))
     }
   }
 
-  def add(priority: Int, hook: () => Unit): AnyRef = synchronized {
+  def add(priority: Int, hook: () => Unit): AnyRef = {
     checkState()
     val hookRef = new SparkShutdownHook(priority, hook)
-    hooks.add(hookRef)
+    hooks synchronized { hooks.add(hookRef) }
```

Comment:
So, technically, this is not really thread-safe anymore. You can have this order of things:

T1: in add(), passes `checkState()` check
T2: in runAll(), sets `shuttingDown` to true
T2: in runAll(), calls `hooks.poll()`, list is empty, stops calling hooks
T1: in add(), adds hook to the list, hook will never be called.

Granted, that's very unlikely. I think if you move the `checkState()` call inside the lock it should fix things, though. You may end up adding a new hook when `shuttingDown` is true in that case, but you make sure that `runAll` will still pick it up.